File: 2022/Day_01/main.hs
import Data.List (groupBy, sortBy)
import Data.Functor ( (<&>) )
import Control.Arrow ( (>>>) )

import Data.List.Split ( splitOn ) -- cabal install --lib split

parseInput :: String -> IO [[Int]]
parseInput filename = readFile filename <&> 
    (splitOn "\n\n" 
    >>> map (lines >>> map read))

part1 :: [[Int]] -> Int
part1 = map sum >>> maximum

part2 :: [[Int]] -> Int
part2 = map sum 
    >>> sortBy (flip compare) 
    >>> take 3
    >>> sum

main :: IO ()
main = do
    elves <- parseInput "input.txt"
    putStrLn $ "Part 1: " ++ show (part1 elves)
    putStrLn $ "Part 2: " ++ show (part2 elves)

    

File: 2022/Day_01/main.exs
defmodule Day_01 do
  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n\n", trim: true)
    |> Enum.map(fn elf ->
      String.split(elf, "\n", trim: true)
      |> Enum.map(&String.to_integer/1)
    end)
  end

  def part1(input) do
    input
    |> Enum.map(&Enum.sum/1)
    |> Enum.max
  end

  def part2(input) do
    input
    |> Enum.map(&Enum.sum/1)
    |> Enum.sort(:desc)
    |> Enum.take(3)
    |> Enum.sum
  end

  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename)
    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2(parsed_input), label: "Part 2")
  end
end

Day_01.main
